nmoscommon/utils.py

```diff
@@ -54,3 +78,82 @@ def getLocalIP():
                 pass
     # Could not find an interface
     return None
+
+
+def downgrade_api_version(
+    obj, rtype, target_ver, downgrade_ver=None, downgrade_map=DOWNGRADE_MAP, remove_keys=["max_api_version"]
+):
+
+    # Sort version list in ascending order
+    version_list = sorted(downgrade_map.keys())
+    # Add 'v1.0' as first element in version list
+    version_list.insert(0, 'v1.0')
+    # Set the max version as last element in list
+    max_ver = version_list[-1]
+
+    # Get current API version, or default to max API version
+    current_api_version = obj.get("@_apiversion", max_ver)
+
+    # Fail if target API version is greater than maximum
+    if api_ver_compare(target_ver, max_ver) > 0:
+        return None
+
+    # Downgrade API object, for a given resource, until it reaches target_version
+    while api_ver_compare(target_ver, current_api_version) < 0:
+        logger.writeDebug("Processing downgrading from {}".format(current_api_version))
+        obj = remove_keys_from_resource(obj, rtype, downgrade_map[current_api_version])
+        current_api_version = version_list[version_list.index(current_api_version) - 1]
+
+    # Set api version key to new downgraded api version
+    if "@_apiversion" in obj:
+        obj["@_apiversion"] = current_api_version
+
+    # Delete any implementation-specific keys not in the spec
+    [obj.pop(key) for key in remove_keys if key in obj]
+
+    # Check if the object's API version is permitted in the output
+    if target_ver == current_api_version:
+        return obj
+    elif downgrade_ver and api_ver_compare(current_api_version, downgrade_ver) >= 0:
```

Comment:
Worth a comment here that this relates to the behaviour of Query API ?query.downgrade=

https://amwa-tv.github.io/nmos-discovery-registration/tags/v1.2.1/docs/2.5._APIs_-_Query_Parameters.html#downgrade-queries